gnomad/utils/constraint.py

```diff
@@ -125,3 +135,246 @@ def collapse_strand(
+            cpg,
+            hl.agg.linreg(
+                observed_variants / possible_variants,
+                [1, mu_snp],
+                weight=possible_variants if weighted else None,
+            ).beta,
+        )
+    }
+    # Build plateau models using sites in population downsamplings if population is specified
+    if pops:
+        pop_lengths = get_all_pop_lengths(ht, pops)
+        for length, pop in pop_lengths:
+            plateau_models_agg_expr[pop] = [
+                hl.agg.group_by(
+                    cpg,
+                    hl.agg.linreg(
+                        ht[f"{prefix}{pop}"][i] / possible_variants,
+                        [1, mu_snp],
+                        weight=possible_variants if weighted else None,
+                    ).beta,
+                )
+                for i in range(length)
+            ]
+
+    return plateau_models_agg_expr
+
+
+def build_coverage_model(
+    low_coverage_obs_exp: hl.Float64Expression,
+    log_coverage: hl.Float64Expression,
+) -> hl.StructExpression:
+    """
+    Build coverage model.
+
+    This function uses linear regression to build a model of log10(coverage) to correct proportion of expected variation at low coverage sites.
+
+    The x and y of the coverage model:
+        - x: log10('exome_coverage') at low coverage site
+        - y: sum('observed_variants')/ (`high_coverage_scale_factor` * sum('possible_variants' * 'mu_snp') at low coverage site
+        where `high_coverage_scale_factor` = sum('observed_variants') / sum('possible_variants' * 'mu_snp') at high coverage site
+
+    :param low_cov_oe_expr: The Float64Expression of observed:expected ratio for a given coverage level.
+    :param log_cov_expr: The Float64Expression of log10 coverage.
+    :return: Tuple with intercept and slope of the model.
+    """
+    return hl.agg.linreg(low_coverage_obs_exp, [1, log_coverage])
+
+
+def get_all_pop_lengths(
+    ht,
+    pops: Tuple[str],
+    prefix: str = "observed_",
+) -> List[Tuple[str, str]]:
+    """
+    Get the minimum length of observed variant counts array for each population downsamping.
+
+    The minimum array length will be used to determine the number of plateau models built for each population.
+
+    The annotations are specified by the combination of `prefix` and each population in `pops`.
+
+    :param ht: Input Table used to build population plateau models.
+    :param pops: Populations used to categorize observed variant counts in downsampings.
+    :param prefix: Prefix of population observed variant counts. Defaults to 'observed_'.
+    :return: A Dictionary with the minimum array length for each population.
+    """
+    pop_downsampling_lengths = ht.aggregate(
+        [hl.agg.min(hl.len(ht[f"{prefix}{pop}"])) for pop in pops]
+    )
+
+    pop_lengths = list(zip(pop_downsampling_lengths, pops))
+    logger.info("Found: %s", "".join(map(str, pop_lengths)))
+
+    for length, pop in pop_lengths:
```

Comment:
I reformat the code so that I can raise errors. Just double check the logic is fine.

Reply:
this seems fine, but I tested it, and it's doing more work; slight preference for reverting and adding a generic error message
<img width="916" alt="image" src="https://user-images.githubusercontent.com/19142091/194621904-f7ebe42f-96a1-4e2b-9251-44f7a660bbb2.png">